### DISCUSSION TITLE
connect ECONNREFUSED 127.0.0.1:80

### DISCUSSION THREAD

**Discussion by imjamesku** [vercel/next.js#13316] [Help] [answered]
In my `getServerSideProps`, I want to make an API call to my next.js app's own API route. the function is as follows:
```
export async function getServerSideProps(context) {
    const searchParams = {
        startDate: moment().subtract(7, 'd').toDate(),
        endDate: moment().toDate(),
        geo: 'US'
    }
    const res = await axios.get(`/api/trends/date?start=${searchParams.startDate.toISOString()}&end=${searchParams.endDate.toISOString()}&geo=${searchParams.geo}`)
    return {
        props: {data: res}
    }
}
```
I get connect ECONNREFUSED 127.0.0.1:80. If I put the data fetching on the client side, it works.  But I'd like to reder the data on the server side. How to I get the SSR to work?

**Comment by jamesmosier** [comment, marked as answer]
Next recommends you write your server code directly in `getServerSideProps`: https://nextjs.org/docs/basic-features/data-fetching#write-server-side-code-directly

Otherwise you'd have to put the fully qualified url in your ```axios.get(`https://my-domain.com/api/trends...`)``` because the server doesn't know what your domain is.

**Reply by imjamesku** [reply]
Thanks. I ended up just using mongoose in getServerSideProps.

**Reply by Suroor-Ahmmad** [reply]
@jamesmosier Hi, need a clarification on the second approach you have mentioned, if the code is yet to be deployed and the api endpoint won't be ready by then, right? So, won't it end up in deployment failure?

**Reply by rafaelnogueira1** [reply]
To use the second approach without problems with deploy for exempla, you can create a .env file and use a variable to store your api url and use that variable here `axios.get(`${process.envMY_VARIABLE_API}/api/trends...`)`.
When you deploy, your server will have the same variables with different values.

**Reply by nickjuntilla** [reply]
This is crazy. You should be able to get data from a server at localhost.

**Reply by m3c-ode** [reply]
Could you be more explicit with the code, in terms of the solution you found?
I wonder hen, what is the point of developing your `/api/` folder?

**Reply by amanbhatt26** [reply]
@m3c-ode I think client side data fetching is a use case for ***/api/*** folder. We can make requests to ***/api/something*** using axios or fetch.

**Reply by sanalpanicker** [reply]
During the build, there is no server running. All server components are prerendered so they can be served during runtime/request time. So the reference to localhost during buildtime isn't going to work unless you have a dedicated server running on that port.

if you want to use the /api route use it with a real host that can serve your data to prerender pages.
/api routes that point to localhost should be used only with client side during runtime/request time ('use client'). At this point you have a server running at localhost that can serve your data.

I think its an issue with your mental model of hoe nextjs works- not an issue with nextjs.

**Comment by alexisLefebvre** [comment]
I had the same error because of a typo, a `/` was missing after `http:/`. If you have the same error, check your URL.

**Reply by Ekundee** [reply]
Thanks so much. You saved my time

**Comment by leoplct** [comment]
In my case I had to replace `http://localhost:1337/api` with `http://127.0.0.1:1337/api`

**Reply by nlszekely-work** [reply]
thanks!!

**Reply by vic-3** [reply]
Thanks, it also worked for me

**Reply by Claudin0305** [reply]
Thanks, it worked for me too

**Reply by colindotfun** [reply]
but *why* does this work 🤔

**Reply by colindotfun** [reply]
also, i still get this error periodically: 

```
Error: connect ECONNREFUSED 127.0.0.1:62596
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1494:16)
    at TCPConnectWrap.callbackTrampoline (node:internal/async_hooks:130:17) {
  errno: -61,
  code: 'ECONNREFUSED',
  syscall: 'connect',
  address: '127.0.0.1',
  port: 62596
}
```

**Reply by TaylanTatli** [reply]
Did you solve your issue? I'm having same problem with latest versions.

**Reply by RutamBhagat** [reply]
@TaylanTatli Same with the Next.js 13.4.1 app router

**Reply by TaylanTatli** [reply]
@RutamBhagat I guess I'm stuck on 13.2.4. 😞 This is the latest working version for me. I don't know what changed. All I'm doing is upgrading packages and it stops working.

**Reply by rinonTen** [reply]
Facing the same issue here. Does anyone have a solution to this problem now?

**Reply by venancioavila** [reply]
same issue: next@13.4.1

**Reply by soophoo** [reply]
Same issue: next@13.2.4

**Reply by VerseVent** [reply]
Same issue: next@13.4.3

**Reply by MoncefDeveloper** [reply]
Same issue: next@13.4.4

**Reply by Tanosimboangy** [reply]
Same issue: Next@13.4.1

**Reply by QuadraKOfx** [reply]
same! Thank you

=> Next@13.4.2

**Reply by matheustodao** [reply]
Same issue: Next@13.4.1

**Reply by ManuelJustMobileSec** [reply]
Same issue next@13.2.4

**Reply by RichardPickman** [reply]
Same  issue: Next@13.3.3. Replace localhost to 127.0.0.1 didn't help

**Reply by dmartin8** [reply]
Same

**Reply by aleksandarrusev** [reply]
Same issue, 13.4.4

**Reply by sanalpanicker** [reply]
During the build, there is no server running. All server components are prerendered so they can be served during runtime/request time. So the reference to localhost during buildtime isn't going to work unless you have a dedicated server running on that port.

if you want to use the /api route use it with a real host that can serve your data to prerender pages.
/api routes that point to localhost should be used only with client side during runtime/request time ('use client'). At this point you have a server running at localhost that can serve your data.

I think its an issue with your mental model of how nextjs works- not an issue with nextjs.

if its not working for you - probably you are using node version less than 17?

**Reply by ajnart** [reply]
Having the same issue. Had to rollback to `v13.2.4` like @TaylanTatli suggested. Works now! Is there a new issue we can follow for this ? It is definitely not fixed @imjamesku

**Reply by Jxck** [reply]
any updates here ?

**Reply by alextenczar** [reply]
Issue still exists on 13.5.4. Reverting to 13.2.4 allowed for the build to execute without an error

**Reply by RaymondMwaura** [reply]
Many thanks sir :triumph:

**Comment by Mergeando** [comment]
I solved it by adding a new path in .env
NEXT_PUBLIC BASE_URL LOCAL=http://127.0.0.1:8000

**Reply by SoolaimanG** [reply]
hello
please how to do resolve this issue

**Reply by ajnart** [reply]
@Mergeando Can you explain more in details ? is the variable name `NEXT_PUBLIC_BASE_URL_LOCAL` ?

**Comment by ashwin-maurya** [comment]
Once check your next version. In my case I was using 13v which was causing the error. Upgraded to 14 and this solved the issue.
